src/pymor/algorithms/rand_la.py

```diff
@@ -2,20 +2,183 @@
+            self.logger.info('Proceeding with maximum operator rank...')
+            basis_size = min(self.A.source.dim, self.A.range.dim)
+        assert 0 < num_testvecs and isinstance(num_testvecs, int)
+        assert 0 < p_fail
+
+        err = self._c_est(num_testvecs, p_fail) * self._maxnorm(basis_size, num_testvecs)
+        self.logger.info(f'estimated error: {err:.10f}')
+
+        return err
+
+    def _extend_basis(self, n=1):
+        self.logger.info(f'Appending {n} basis vector{"s" if n > 1 else ""}.')
+
+        with self._basis_rng_real:
+            W = self.A.source.random(n, distribution='normal')
+        if self.complex:
+            with self._basis_rng_imag:
+                W += 1j * self.A.source.random(n, distribution='normal')
+
+        self._Q[0].append(self.A.apply(W))
+        gram_schmidt(self._Q[0], self.range_product, offset=self._l, copy=False)
+
+        for i in range(self.subspace_iterations):
+            i = 2*i + 1
+            self._Q[i].append(self.source_product.apply_inverse(
+                (self.A.apply_adjoint(self.range_product.apply(self._Q[i-1][-n:])))))
+            gram_schmidt(self._Q[i], self.source_product, offset=self._l, copy=False)
+            self._Q[i+1].append(self.A.apply(self._Q[i][-n:]))
+            gram_schmidt(self._Q[i+1], self.range_product, offset=self._l, copy=False)
+
+        self._l += n
+
+    def _find_range(self, basis_size=8, tol=None, num_testvecs=20, p_fail=1e-14, block_size=8, increase_block=True,
+                    max_basis_size=500):
+        if basis_size > self._l:
+            self._extend_basis(basis_size - self._l)
+
+        if tol is not None and self.estimate_error(basis_size, num_testvecs, p_fail) > tol:
+            with self.logger.block('Extending range basis adaptively...'):
+                max_iter = min(max_basis_size, self.A.source.dim, self.A.range.dim)
+                while self._l < max_iter:
+                    if increase_block:
+                        low = basis_size
+                        basis_size += block_size
+                        block_size *= 2
+                    else:
+                        basis_size += block_size
+                    basis_size = min(basis_size, max_iter)
+                    if self.estimate_error(basis_size, num_testvecs, p_fail) <= tol:
+                        break
+            if increase_block:
+                with self.logger.block('Contracting range basis...'):
+                    high = basis_size
+                    while True:
+                        mid = high - (high - low) // 2
+                        if basis_size == mid:
+                            break
+                        basis_size = mid
+                        err = self.estimate_error(basis_size, num_testvecs, p_fail)
+                        if err <= tol:
+                            high = mid
+                        else:
+                            low = mid
```

Comment:
I guess by adaptivity you mean increasing `blocksize`? 
I implemented this in case there is a benefit in the `gram_schmidt` method from calling it in a blocked fashion and if not (which seems the case) maybe discuss about modifying it ;-)

Reply:
Yeah, in it's current implementation `gram_schmidt` is strictly sequential regarding the vectors to orthonormalize. A blocked version would come at the expense of numerical accuracy, if I'm not mistaken. In some use cases, that might be ok, of course.

Reply:
OK. I have removed that feature for now. I might come back to that later in a seperate PR that also adds blocked Gram Schmidt.